File: LeetCode/Python/Easy/404-Sum-of-Left-Leaves/solution.py
class Solution:
    def sumOfLeftLeaves(self, root: Optional[TreeNode]) -> int:
        (tree_flag, result)=self.in_order_traversal(root, False)

        return result

    def in_order_traversal(self, root: Optional[TreeNode], came_from_left: bool) -> (bool, int):
        if not root:
            return (False, 0)
        
        if not root.left and not root.right:
            if came_from_left:
                return (True, root.val)
            else:
                return (False, 0)

        subtree_total=0

        subtree_flag=False

        left_subtree_flag, left_subtree_total=self.in_order_traversal(root.left, True)

        right_subtree_flag, right_subtree_total=self.in_order_traversal(root.right, False)

        if left_subtree_flag:
            subtree_total+=left_subtree_total

            subtree_flag=True

        if right_subtree_flag:
            subtree_total+=right_subtree_total

            subtree_flag=True

        return (subtree_flag, subtree_total)

class Solution:
    def sumOfLeftLeaves(self, root: Optional[TreeNode]) -> int:
        result=0

        stack=list()

        stack.append((False, root))

        while stack:
            came_from_left, parent=stack.pop()

            if parent.left==None and parent.right==None:
                if came_from_left:
                    result+=parent.val

                continue

            if parent.right:
                stack.append((False, parent.right))

            if parent.left:
                stack.append((True, parent.left))

        return result

File: LeetCode/Rust/Easy/404-Sum-of-Left-Leaves/solution.rs
use std::rc::Rc;
use std::cell::{Ref, RefCell, RefMut};
impl Solution
{
    pub fn sum_of_left_leaves(root: Option<Rc<RefCell<TreeNode>>>) -> i32
    {
        let (tree_flag, result): (bool, i32)=Solution::in_order_traversal(root.as_ref(), false);

        result
    }

    fn in_order_traversal(root: Option<&Rc<RefCell<TreeNode>>>, came_from_left: bool) -> (bool, i32)
    {
        match root
        {
            Some(root_rc) =>
            {
                let node: Ref<'_, TreeNode>=root_rc.borrow();

                if node.left.is_none() && node.right.is_none()
                {
                    if came_from_left
                    {
                        return (true, node.val);
                    }
                    else
                    {
                        return (false, 0);
                    }
                }

                let mut subtree_total: i32=0;

                let mut subtree_flag: bool=false;

                let (left_subtree_flag, left_subtree_total): (bool, i32)=Solution::in_order_traversal(node.left.as_ref(), true);

                let (right_subtree_flag, right_subtree_total): (bool, i32)=Solution::in_order_traversal(node.right.as_ref(), false);

                if left_subtree_flag
                {
                    subtree_total+=left_subtree_total;

                    subtree_flag=true;
                }

                if right_subtree_flag
                {
                    subtree_total+=right_subtree_total;

                    subtree_flag=true;
                }

                (subtree_flag, subtree_total)
            },
            None => (false, 0)
        }
    }
}

impl Solution
{
    pub fn sum_of_left_leaves(root: Option<Rc<RefCell<TreeNode>>>) -> i32
    {
        let mut result: i32=0;

        let mut stack: Vec<(Option<Rc<RefCell<TreeNode>>>, bool)>=Vec::new();

        stack.push((root, false));

        while let Some((parent_option, came_from_left))=stack.pop()
        {
            match parent_option
            {
                Some(parent_rc) =>
                {   
                    let mut parent: RefMut<'_, TreeNode>=parent_rc.borrow_mut();

                    match (parent.left.take(), parent.right.take())
                    {
                        (Some(left_child), Some(right_child)) =>
                        {
                            stack.push((Some(left_child), true));

                            stack.push((Some(right_child), false));
                        },
                        (Some(left_child), None) =>
                        {
                            stack.push((Some(left_child), true));
                        },
                        (None, Some(right_child)) =>
                        {
                            stack.push((Some(right_child), false));
                        },
                        (None, None) =>
                        {
                            if came_from_left
                            {
                                result+=parent.val;
                            }
                        },
                        _ => ()
                    }
                },
                None => continue
            }
        }

        result
    }
}

impl Solution
{
    pub fn sum_of_left_leaves(root: Option<Rc<RefCell<TreeNode>>>) -> i32
    {
        let mut result: i32=0;

        let mut stack: Vec<(Option<Rc<RefCell<TreeNode>>>, bool)>=Vec::new();

        stack.push((root, false));

        while let Some((parent_option, came_from_left))=stack.pop()
        {
            match parent_option
            {
                Some(parent_rc) =>
                {   
                    let parent: Ref<'_, TreeNode>=parent_rc.borrow();

                    match (parent.left.clone(), parent.right.clone())
                    {
                        (Some(left_child), Some(right_child)) =>
                        {
                            stack.push((parent.left.clone(), true));

                            stack.push((parent.right.clone(), false));
                        },
                        (Some(left_child), None) =>
                        {
                            stack.push((parent.left.clone(), true));
                        },
                        (None, Some(right_child)) =>
                        {
                            stack.push((parent.right.clone(), false));
                        },
                        (None, None) =>
                        {
                            if came_from_left
                            {
                                result+=parent.val;
                            }
                        },
                        _ => ()
                    }
                },
                None => continue
            }
        }

        result
    }
}